### PR TITLE
Update github.com/imdario/mergo calls from mergo.Merge to mergo.MergeWithOverwrite

### PR DESCRIPTION
**Pull request by bufdev**
Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/12272
Also per https://github.com/imdario/mergo/releases/tag/0.2.0

CLA signed and passes unit tests in `pkg/client/...`:

```
$ cd pkg/client && godep go test ./...
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client    0.353s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache  0.242s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/chaosclient    0.006s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd  0.027s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api  0.017s
?       github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api/latest   [no test files]
?       github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api/v1   [no test files]
?       github.com/GoogleCloudPlatform/kubernetes/pkg/client/metrics    [no test files]
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/portforward    0.016s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/record 0.057s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/remotecommand  0.034s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/testclient 0.012s
```

### CODE REVIEW

**Comment by peter-edge** [comment]
I could be wrong but I don't think the Shippable failure is real

**Comment by dchen1107** [comment]
ok to test

**Comment by dchen1107** [comment]
cc/ @deads2k

File: pkg/client/clientcmd/loader.go [hunk 1]
```diff
@@ -134,8 +134,8 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
 	// since values are overwritten, but maps values are not, we can merge the non-map config on top of the map config and
 	// get the values we expect.
 	config := clientcmdapi.NewConfig()
-	mergo.Merge(config, mapConfig)
-	mergo.Merge(config, nonMapConfig)
+	mergo.MergeWithOverwrite(config, mapConfig)
```
Comment (deads2k) [inline pkg/client/clientcmd/loader.go]: Previously, map values were only overwritten if no key was set in the destination map before.  With this change, the destination map value will be overwritten.  I think the new semantic could be used to make this function easier to ready by eliminating the distinction between `mapConfig` and `nonMapConfig`, but as it is now isn't the `mapConfig` essentially stomped?

File: pkg/client/clientcmd/client_config.go [hunk 2]
```diff
@@ -109,13 +109,13 @@ func (config DirectClientConfig) ClientConfig() (*client.Config, error) {
 		if err != nil {
 			return nil, err
 		}
-		mergo.Merge(clientConfig, userAuthPartialConfig)
+		mergo.MergeWithOverwrite(clientConfig, userAuthPartialConfig)
```
Comment (deads2k) [inline pkg/client/clientcmd/client_config.go]: You'll need to update the comments a couple lines up.  They refer to the old merge semantics.

File: pkg/client/clientcmd/loader.go (hunk 1 above)
Comment (peter-edge) [inline pkg/client/clientcmd/loader.go]: This isn't completely accurate - previously, values would be stomped if a new non-zero value was set in object being merged. That this changed to no overwrite is what caused https://github.com/GoogleCloudPlatform/kubernetes/issues/12272.

I can look at this specific code in a bit but the intention here was just to maintain existing functionality.

File: pkg/client/clientcmd/client_config.go (hunk 2 above)
Comment (peter-edge) [inline pkg/client/clientcmd/client_config.go]: will do!

File: pkg/client/clientcmd/loader.go (hunk 1 above)
Comment (deads2k) [inline pkg/client/clientcmd/loader.go]: > This isn't completely accurate - previously, values would be stomped if a new non-zero value was set in object being merged. That this changed to no overwrite is what caused #12272.

Map values were first write wins, not last write wins.  Struct values were last write wins.  `MergeWithOverwrite` makes the map values last write wins too.  Hence all the comments explaining why multiple objects were merged in reverse orders.

Comment (peter-edge) [inline pkg/client/clientcmd/loader.go]: ah noted. ok I will check this all out in a bit and update, sorry bout that

**Comment by peter-edge** [comment]
@deads2k another option i hate to mention is we have kubernetes' own version of mergo in pkg/ and not rely on an external library. for something like this, it might not be a bad idea.

**Comment by deads2k** [comment]
> @deads2k another option i hate to mention is we have kubernetes' own version of mergo in pkg/ and not rely on an external library. for something like this, it might not be a bad idea.

I don't see anything particularly wrong with using this external library.  I'm just pretty sure that semantics changed (for the better I think) and we'll update with it.

**Comment by peter-edge** [comment]
Closing this for now and will redo in another PR - had a git snafu locally and will be easier to just do the substitute/replace in a new PR.

**Pull request closed without merging**